synapse/lib/stormlib/infosec.py

```diff
@@ -50,6 +51,450 @@
+    ModifiedIntegrity = COEFFS['MI']
+    AvailabilityRequirement = COEFFS['AR']
+    ModifiedAvailability = COEFFS['MA']
+
+    MISS = min(
+        1.0 - ((1.0 - ConfidentialityRequirement * ModifiedConfidentiality) *
+                (1.0 - IntegrityRequirement * ModifiedIntegrity) *
+                (1.0 - AvailabilityRequirement * ModifiedAvailability)
+        ),
+        0.915
+    )
+
+    if _ModifiedScope == 'U':
+        ModifiedImpact = 6.42 * MISS
+    else:
+        ModifiedImpact = 7.52 * (MISS - 0.029) - 3.25 * (MISS * 0.9731 - 0.02) ** 13
+
+    ModifiedAttackVector = COEFFS['MAV']
+    ModifiedAttackComplexity = COEFFS['MAC']
+    ModifiedPrivilegesRequired = COEFFS['MPR']
+    ModifiedUserInteraction = COEFFS['MUI']
+
+    ModifiedExploitability = (
+        8.22 *
+        ModifiedAttackVector *
+        ModifiedAttackComplexity *
+        ModifiedPrivilegesRequired *
+        ModifiedUserInteraction
+    )
+
+    if ModifiedImpact <= 0:
+        EnvironmentalScore = 0.0
+    else:
+        ExploitCodeMaturity = COEFFS['E']
+        RemediationLevel = COEFFS['RL']
+        ReportConfidence = COEFFS['RC']
+
+        if _ModifiedScope == 'U':
+            EnvironmentalScore = ROUND(
+                ROUND(
+                    min(ModifiedImpact + ModifiedExploitability, 10)) *
+                ExploitCodeMaturity *
+                RemediationLevel *
+                ReportConfidence
+            )
+
+        else:
+            EnvironmentalScore = ROUND(
+                ROUND(
+                    min(1.08 * (ModifiedImpact + ModifiedExploitability), 10)) *
+                ExploitCodeMaturity *
+                RemediationLevel *
+                ReportConfidence
+            )
+
+    return BaseScore, TemporalScore, EnvironmentalScore
+
+def CVSS3_0_calc(vdict):
+    return _CVSS3_calc(vdict, CVSS3_0_STR)
+
+def CVSS3_1_calc(vdict):
+    return _CVSS3_calc(vdict, CVSS3_1_STR)
+
+CVSS_CALC = {
+    CVSS2_STR: CVSS2_calc,
+    CVSS3_0_STR: CVSS3_0_calc,
+    CVSS3_1_STR: CVSS3_1_calc,
+}
+
+def validate(vect, vers):
+    '''
+    Validate (as best as possible) the vector string. Look for issues such as:
+        - No duplicated metrics
+        - Invalid metrics
+        - Invalid metric values
+        - Missing mandatory metrics
+
+    Returns a dictionary with the parsed metric:value pairs.
+    '''
+
+    missing = []
+    badvals = []
+    invalid = []
+
+    TAG = CVSS_TAGS[vers]
+    METRICS = CVSS_METRICS[vers]
+
+    # Do some canonicalization of the vector for easier parsing
+    vect = vect
+    vect = vect.strip('(')
+    vect = vect.strip(')')
+
+    if vect.startswith(TAG):
+        vect = vect[len(TAG):]
+
+    try:
+        # Parse out metrics
+        mets_vals = [k.split(':') for k in vect.split('/')]
+
+        # Convert metrics into a dictionary
+        vdict = dict(mets_vals)
+
+    except ValueError:
+        raise s_exc.BadDataValu(mesg='Malformed vector string')
```

Comment:
`f`Malformed CVSS vector: {vect}'` for user clarity since these types of errors occasionally bubble up from within power-ups etc. :)

Reply:
Fixed.